cpp/src/arrow/util/bitmap_ops.h

```diff
@@ -147,6 +147,31 @@ bool OptionalBitmapEquals(const std::shared_ptr<Buffer>& left, int64_t left_offs
                           const std::shared_ptr<Buffer>& right, int64_t right_offset,
                           int64_t length);
 
+/// \brief Do a "bitmap and" on right and left buffers starting at
+/// their respective bit-offsets for the given bit-length and put
+/// the results in out_buffer starting at the given bit-offset.
+/// Both right and left buffers are optional. If any of the buffers is
+/// null a bitmap of zeros is returned.
```

Comment:
Hmm, actually, the more useful semantics is that a null pointer means the bitmap is all-1s.

This reflects the situation where an Array doesn't have a null bitmap: all values are valid.

Reply:
And this would be even more useful as:
```c++
Result<std::shared_ptr<Buffer>> OptionalBitmapAnd(
    MemoryPool* pool, const std::shared_ptr<Buffer>& left, int64_t left_offset,
    const std::shared_ptr<Buffer>& right, int64_t right_offset);
```

... because then, if one of the inputs is null, and the offsets are compatible, we can return the other input (perhaps sliced) instead of allocating a new buffer.